### PR TITLE
Allow deploy keys as alternative to access tokens

### PR DESCRIPTION
**Pull request by Trundle**
This pull request makes the action also work with deploy keys instead of access tokens. Compared to a personal access token, deploy keys have the advantage that they can be limited to a single repository.

### CODE REVIEW

**Comment by SeanTAllen** [comment]
@Trundle- you can use `make pylint` to run the linting that will happen at CI time locally. As long as you have docker installed it will work.

File: Dockerfile [hunk 1]
```diff
@@ -12,7 +12,8 @@ RUN apk add --update --no-cache \
   make \
   python3 \
   python3-dev \
-  py3-pip
+  py3-pip \
+  openssh-client
```
Comment (SeanTAllen) [inline Dockerfile:16]: can you put this in alphabetical order before python3?

File: entrypoint.py [hunk 2]
```diff
@@ -19,8 +23,10 @@
 INFO = '\033[34m'
 NOTICE = '\033[33m'
 
-if 'RELEASE_TOKEN' not in os.environ:
-    print(ERROR + "RELEASE_TOKEN needs to be set in env. Exiting." + ENDC)
+deploy_key = os.environ.get('DEPLOY_KEY')
+if not deploy_key and 'RELEASE_TOKEN' not in os.environ:
+    print(f"{ERROR}Either RELEASE_TOKEN or DEPLOY_KEY needs to be set in env. "
+          f"Exiting.{ENDC}")
```
Comment (SeanTAllen) [inline entrypoint.py:29]: I believe you can do a `if (X,Y) not in os.environ` here.

Like:

```python
a = [1,2,3,4]

if 1 or 5 in a:
  print("hello")
```

**Comment by SeanTAllen** [comment]
@Trundle how much of this have you tested and how? 

Normally I test in the action-testing repo and verify everything works (but you arent a contributor to that- so I could test it all there).

It took a while to get the "retry on failure" code working so I'm wary of changes to it.

File: README.md [hunk 3]
```diff
@@ -38,6 +38,8 @@ jobs:
 
 N.B. The environment variable RELEASE_TOKEN that is required by each step must be a personal access token with public_repo access. You can not use the GITHUB_TOKEN environment variable provided by GitHub's action environment. If you try to use GITHUB_TOKEN, the action will fail when trying to upload the built documentation.
 
+Alternatively, you can use a [Deploy Key](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) with write access by setting the `DEPLOY_KEY` environment variable.
+
```
Comment (SeanTAllen) [inline README.md:42]: Can you update the README to show an example that uses the deploy key and not a release token?

**Comment by Trundle** [comment]
I tested it in https://github.com/Trundle/pony-templates, pointing to my fork of the action: https://github.com/Trundle/pony-templates/runs/2171361022?check_suite_focus=true

I only tested with a deploy key though.

**Comment by SeanTAllen** [comment]
@EpicEric can you give this a review as well?

**Comment by SeanTAllen** [comment]
@Trundle I'll give it a go in the action-testing as well now.

File: entrypoint.py [hunk 4]
```diff
@@ -233,23 +239,62 @@
...
+    # That's fine, it will exist after our push. Just not the error and move on.
+    try:
+        git.reset('gh-token/generated-documentation')
+    except GitCommandError:
+        print(NOTICE + "Couldn't git reset generated-documentation." + ENDC)
+        print(NOTICE + "This error is expected if the branch doesn't exist yet."
+          + ENDC)
+
+    print(INFO + "Running 'mkdocs gh-deploy'." + ENDC)
+    # pylint: disable=W1510
```
Comment (SeanTAllen) [inline entrypoint.py:292]: is this related to the `cwd` change?

what made this necessary? 

I think that an explicit: os.chdir(docs_build_dir) is more clear. and more likely to give a good error message, but I'm not sure about that.

**Comment by SeanTAllen** [comment]
I tested with existing config in action-testing and it all worked, so the happy path is good. Still a little wary of the change in the git push failure. I'd like @EpicEric to review as he is better with python than I am.

But it looks pretty good other than a few things here and there. Thanks @Trundle. Ping me when you've made some of the requested updates.

File: entrypoint.py (hunk 2 above)
Comment (Trundle) [reply entrypoint.py:29]: `,` constructs tuples in Python, so `(X,Y) not in os.environ` will check whether the tuple `x, y` exists as key in `os.environ` and will always be false.

In your example, you are checking whether `1` is true or whether `5` is contained in `a`:

```pycon
>>> a = [2, 3, 4]
>>> 1 or 5 in a
1
>>> bool(1 or 5 in a)
True
>>> if 1 or 5 in a: print('whoops')
... 
whoops
```

What the code above does, maybe a bit too subtle: it sets the value of env var `DEPLOY_KEY` to the name `deploy_key`. If no such env var exists, the value will be `None`. It then checks whether `deploy_key` evaluates to false. That is the case for `None` as well as an empty string, so if you set the env var `DEPLOY_TOKEN`, but to an empty string, it will be caught here as well. I can change that to a bit more explicit code though if you like.

Comment (SeanTAllen) [reply entrypoint.py:29]: I think that would be good. I expect we will have people who are not very python knowledgeable wanting to contribute (like me!) so more explicit is better.

File: entrypoint.py (hunk 4 above)
Comment (Trundle) [reply entrypoint.py:292]: It suppresses the following pylint warning:

```
entrypoint.py:286:11: W1510: Using subprocess.run without explicitly set `check` is not recommended. (subprocess-run-check)
```

`check=True` will raise an exception if the process exits with status != 0. As we always check the status code, that's not an issue. Plus I think an exception in the output would be way more confusing than the error message we print instead.

As I'm not using any of subprocess's features anyway, I can also change the code back to `os.chdir` and `os.system`. I initially wanted to pass a modified environment to the subprocess, hence the change to the subprocess module.

Comment (SeanTAllen) [reply entrypoint.py:292]: You know Python better than I, that's clear. Whatever you think is the easiest to maintain and for someone like me to understand, let's do that.

File: entrypoint.py (hunk 2 above)
Comment (SeanTAllen) [inline entrypoint.py:29]: If you are going to change the prints to using f"" for errors, can you update them all in one go so there's a common pattern to them?

**Comment by ponylang-main** [comment]
Hi @Trundle,

The changelog - added label was added to this pull request; all PRs with a changelog label need to have release notes included as part of the PR. If you haven't added release notes already, please do.

Release notes are added by creating a uniquely named file in the `.release-notes` directory. We suggest you call the file `13.md` to match the number of this pull request.

The basic format of the release notes (using markdown) should be:

```
## Title

End user description of changes, why it's important,
problems it solves etc.

If a breaking change, make sure to include 1 or more
examples what code would look like prior to this change
and how to update it to work after this change.
```

Thanks.

File: entrypoint.py (hunk 2 above)
Comment (Trundle) [reply entrypoint.py:29]: :+1: A common pattern is worth having. I can also replace the f strings to string concatenation, do you have any preference? Personally, I have a slight tendency to f strings as they are a bit more compact, but that is all a question of preference and habit. I can also change everything to f strings.

Comment (SeanTAllen) [reply entrypoint.py:29]: I don't much care as long as it's consistent within any particular grouping of things.

File: entrypoint.py [hunk 5]
```diff
@@ -233,23 +239,62 @@
...
+                        f'-i {identity_file.name} $@')
+                    os.chmod(ssh_wrapper_path, 0o500)
+
+                identity_file.write(deploy_key.encode('utf-8'))
+                identity_file.flush()
+                os.environ['GIT_SSH'] = ssh_wrapper_path
+                try:
+                    yield
+                finally:
+                    del os.environ['GIT_SSH']
```
Comment (EpicEric) [inline entrypoint.py:264]: It's probably better to use [`custom_environment()`](https://gitpython.readthedocs.io/en/stable/reference.html#git.cmd.Git.custom_environment) here.

Comment (Trundle) [reply entrypoint.py:264]: That's what I initially though as well, but the `mkdocs` invocation will also require Git credentials. That's why I went with this solution instead.

**Review by EpicEric — APPROVED** [review]

**Comment by SeanTAllen** [comment]
@Trundle is this ready to squash and merge?

**Review by SeanTAllen — APPROVED** [review]

**Comment by Trundle** [comment]
So it stopped working in my repository now, but I don't see why yet :raised_eyebrow: I think I messed up my keys. I'll have a look tomorrow.

**Comment by SeanTAllen** [comment]
@Trundle ok. let me know if I should re-test any functionality.

How did you determine that it stopped working? Did you do a run and it failed at some point?

**Comment by Trundle** [comment]
Exactly. I triggered a run (https://github.com/Trundle/pony-templates/runs/2180017638?check_suite_focus=true) and it fails when fetching and complains about SSH keys. I deleted my previous deploy keys, hence I think this is rather some mistake in my setup, but I'm too tired right now to think it through.

**Comment by Trundle** [comment]
I learned that OpenSSH requires keys to end with a newline. I now added an explicit check whether the passed key ends in a newline, because I expect that to be a common c/p mistake with GitHub secrets.

There was also a refactoring glitch when I changed back to `os.system`, fixed with 2721b94bbde3d97e685e2218d8bef904204a6745. I think this is good to squash and merge now!

**Comment by SeanTAllen** [comment]
Thanks @Trundle. I'll give it another quick test and then... Away we go!
